### PR TITLE
Add pysam and rust-bio examples

### LINKED ISSUE #1: Suggestions for fastq benchmark
Sorry if I have jumped the gun here and these suggestions are forthcoming. These would be the two libraries I use most (and I suspect quite a few other do too) for the respective languages. I'm interested to see where they fall in the existing benchmark.

- Python [`pysam.FastxFile`](https://pysam.readthedocs.io/en/latest/api.html#pysam.FastxFile) from [pysam][pysam].
- Rust [`bio::io::fastq::Reader`](https://docs.rs/bio/0.30.1/bio/io/fastq/struct.Reader.html) from the [rust-bio][rust] library

Not sure if this is also of interest, but [hyperfine][hyperfine] is a nice tool for doing this kind of benchmarking. Not only will it give relative times with multiple runs, but also allows for warm-up runs (useful for IO intensive processes). I've used it for a very similar task [here](https://github.com/mbhall88/rasusa#benchmark).

**EDIT:**  
Just read the accompanying blog again and noticed you mentioned you implemented in languages you are familiar with. I'm happy to provide any Rust examples if needed.

[pysam]: https://pysam.readthedocs.io/en/latest/
[rust-bio]: https://rust-bio.github.io/
[hyperfine]: https://github.com/sharkdp/hyperfine

### PR DESCRIPTION
**Pull request by mbhall88**
Closes #1 

## Added
-  A pysam implementation for Python
- A rust-bio implementation for Rust

## Changed
- `Makefile` now has rules for compiling the Rust binary `make rust` (or `make fqcnt_rs1`)

# Benchmark
I ran the C, Go, Rust, Pysam, and pyfastx implementations with [hyperfine](https://github.com/sharkdp/hyperfine) as follows

```sh
cd fqcnt/
fq="../data/M_abscessus_HiSeq.fq"
C_CMD="./fqcnt_c1_kseq $fq"
RS_CMD="./bin/fqcnt_rustbio $fq"
GO_CMD="./fqcnt_go1 $fq"
PYFASTX_CMD="./fqcnt_py6x_pyfx.py $fq"
PYSAM_CMD="./fqcnt_py7x_pysam.py $fq"
hyperfine --warmup 3 --runs 10 --export-markdown results.md \
  "$C_CMD" "$RS_CMD" "$GO_CMD" "$PYFASTX_CMD" "$PYSAM_CMD"
```

Note: This was run on my cluster (which has been having some major IO slowdowns recently) so the environment is definitely not as isolated as the one you use in your blog @lh3. I suspect these times would be much lower in your environment.

### Non-compressed

| Command | Mean [s] | Min…Max [s] |
|:---|---:|---:|
| `./fqcnt_c1_kseq ../data/M_abscessus_HiSeq.fq` | 1.054 ± 0.014 | 1.044…1.092 |
| `./bin/fqcnt_rustbio ../data/M_abscessus_HiSeq.fq` | 2.494 ± 0.046 | 2.456…2.598 |
| `./fqcnt_go1 ../data/M_abscessus_HiSeq.fq` | 2.221 ± 0.122 | 2.015…2.436 |
| `./fqcnt_py6x_pyfx.py ../data/M_abscessus_HiSeq.fq` | 6.885 ± 0.210 | 6.573…7.248 |
| `./fqcnt_py7x_pysam.py ../data/M_abscessus_HiSeq.fq` | 12.035 ± 0.661 | 11.332…13.409 |

```
Summary
  './fqcnt_c1_kseq ../data/M_abscessus_HiSeq.fq' ran
    2.11 ± 0.12 times faster than './fqcnt_go1 ../data/M_abscessus_HiSeq.fq'
    2.37 ± 0.05 times faster than './bin/fqcnt_rustbio ../data/M_abscessus_HiSeq.fq'
    6.53 ± 0.22 times faster than './fqcnt_py6x_pyfx.py ../data/M_abscessus_HiSeq.fq'
   11.42 ± 0.64 times faster than './fqcnt_py7x_pysam.py ../data/M_abscessus_HiSeq.fq'
```

### Compressed

| Command | Mean [s] | Min…Max [s] |
|:---|---:|---:|
| `./fqcnt_c1_kseq ../data/M_abscessus_HiSeq.fq.gz` | 8.441 ± 0.033 | 8.410…8.495 |
| `./bin/fqcnt_rustbio ../data/M_abscessus_HiSeq.fq.gz` | 10.667 ± 0.034 | 10.629…10.753 |
| `./fqcnt_go1 ../data/M_abscessus_HiSeq.fq.gz` | 17.187 ± 0.310 | 16.822…17.699 |
| `./fqcnt_py6x_pyfx.py ../data/M_abscessus_HiSeq.fq.gz` | 14.038 ± 0.227 | 13.754…14.454 |
| `./fqcnt_py7x_pysam.py ../data/M_abscessus_HiSeq.fq.gz` | 17.721 ± 0.405 | 17.209…18.552 |

```
Summary
  './fqcnt_c1_kseq ../data/M_abscessus_HiSeq.fq.gz' ran
    1.26 ± 0.01 times faster than './bin/fqcnt_rustbio ../data/M_abscessus_HiSeq.fq.gz'
    1.66 ± 0.03 times faster than './fqcnt_py6x_pyfx.py ../data/M_abscessus_HiSeq.fq.gz'
    2.04 ± 0.04 times faster than './fqcnt_go1 ../data/M_abscessus_HiSeq.fq.gz'
    2.10 ± 0.05 times faster than './fqcnt_py7x_pysam.py ../data/M_abscessus_HiSeq.fq.gz'
```

I ran the exact same commands, just using the `gzip` compressed version of th fastq file.

### CODE REVIEW

File: fqcnt/fqcnt_rs1_rustbio.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    let records = fastq::Reader::new(stream).records().map(|r| r.unwrap());
+    let mut n: u32 = 0;
+    let mut slen: u64 = 0;
+    let mut qlen: u64 = 0;
+    for record in records {
+        n += 1;
+        slen += record.seq().len() as u64;
+        qlen += record.qual().len() as u64;
+    }
```
Comment (dlaehnemann) [inline fqcnt/fqcnt_rs1_rustbio.rs:65]: @mbhall88 Maybe using a pre-allocated mutable record might further improve the timing a bit. Would you mind trying that out with your setup?

For an example of this setup, checkout the FastqRead Trait docs: https://docs.rs/bio/0.30.1/bio/io/fastq/trait.FastqRead.html

**Comment by lh3** [comment]
@dlaehnemann and @mbhall88 are you interested in providing a rust implementation for the bedcov benchmark? It is probably more informative about the capability of the language itself. This algorithm can be useful to rust-bio, too (rust-bio has interval_tree.rs, but static index data structures like cgranges are going to be much faster). The core algorithm is described in this [C++ header](https://github.com/lh3/cgranges/blob/master/cpp/IITree.h). It is short.

File: fqcnt/fqcnt_rs1_rustbio.rs (hunk 1 above)
Comment (mbhall88) [reply fqcnt/fqcnt_rs1_rustbio.rs:65]: Not sure I follow sorry? Maybe you could illustrate with a PR?

**Comment by mbhall88** [comment]
> @dlaehnemann and @mbhall88 are you interested in providing a rust implementation for the bedcov benchmark? It is probably more informative about the capability of the language itself. This algorithm can be useful to rust-bio, too (rust-bio has interval_tree.rs, but static index data structures like cgranges are going to be much faster). The core algorithm is described in this [C++ header](https://github.com/lh3/cgranges/blob/master/cpp/IITree.h). It is short.

I would be happy to have a go. Two things, however: it sounds like @dlaehnemann is much more accomplished at Rust than myself, and I cannot do it right now (may need a week or so).  

Either way, if no one has tackled this in the next week or so I will put something together.

File: fqcnt/fqcnt_rs1_rustbio.rs (hunk 1 above)
Comment (dlaehnemann) [reply fqcnt/fqcnt_rs1_rustbio.rs:65]: Yeah, sorry, that was jotted down a bit shoddily. I meant something like this:
```
    let mut reader = fastq::Reader::new(stream);
    let mut record = fastq::Record::new();

    let mut n: u32 = 0;
    let mut slen: u64 = 0;
    let mut qlen: u64 = 0;

    reader.read(&mut record)?;
    while !record.is_empty() {
        n += 1;
        slen += record.seq().len() as u64;
        qlen += record.qual().len() as u64;
        reader.read(&mut record)?
    }
```
A colleague of mine (@tedil) tested this locally and cut the time further. According to @johanneskoester it avoids re-allocating a new record on every iteration.

**Comment by dlaehnemann** [comment]
I'd pass that compliment right back: I'm quite sure you are more accomplished, here. But I'll be glad to collaborate on this and could try to pull in other people for help and review. ;) @tedil @johanneskoester

So maybe just cross-post here or mention me elsewhere, once you have started on something. Should I start something first, I'll do the same.

File: fqcnt/fqcnt_rs1_rustbio.rs (hunk 1 above)
Comment (mbhall88) [reply fqcnt/fqcnt_rs1_rustbio.rs:65]: Oh, cool! DIdn't realise that would have such a dramatic effect. For some reason, I assumed that Rust's Iterators would have just implicitly done the same thing when looping over them.

Comment (mbhall88) [reply fqcnt/fqcnt_rs1_rustbio.rs:65]: https://github.com/lh3/biofast/pull/12

**Pull request merged**